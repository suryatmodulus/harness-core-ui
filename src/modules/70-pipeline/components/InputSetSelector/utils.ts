import type { IconName, SelectOption } from '@harness/uicore'
import type { EntityGitDetails, InputSetSummaryResponse } from 'services/pipeline-ng'
import css from './InputSetSelector.module.scss'

type InputSetLocal = InputSetSummaryResponse & SelectOption
export interface InputSetValue extends InputSetLocal {
  type: InputSetSummaryResponse['inputSetType']
  gitDetails?: EntityGitDetails
}

export const getIconByType = (type: InputSetSummaryResponse['inputSetType']): IconName => {
  return type === 'OVERLAY_INPUT_SET' ? 'step-group' : 'yaml-builder-input-sets'
}

export const onDragStart = (event: React.DragEvent<HTMLLIElement>, row: InputSetValue): void => {
  event.dataTransfer.setData('data', JSON.stringify(row))
  event.currentTarget.classList.add(css.dragging)
}

export const onDragEnd = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragging)
}

export const onDragLeave = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragOver)
}

export const onDragOver = (event: React.DragEvent<HTMLLIElement>): void => {
  if (event.preventDefault) {
    event.preventDefault()
  }
  event.currentTarget.classList.add(css.dragOver)
  event.dataTransfer.dropEffect = 'move'
}
