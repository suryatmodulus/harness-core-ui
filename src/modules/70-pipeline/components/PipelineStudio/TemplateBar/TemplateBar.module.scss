/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.container {
  border-color: transparent !important;
  .actionButton {
    transform: rotate(90deg);
    cursor: pointer;
  }
}

.main {
  position: relative;
  :global {
    .bp3-overlay {
      display: block !important;
    }
    .bp3-overlay,
    .bp3-transition-container {
      width: 100%;
    }
  }
}
.popover {
  background-color: var(--red);
  :global {
    .bp3-popover {
      width: 100%;
      box-shadow: 0px 0px 1px rgb(40 41 61 / 4%), 0px 2px 4px rgb(96 97 112 / 16%);
      border: 0.5px solid rgba(0, 0, 0, 0.1);
      border-radius: var(--spacing-2);
      transform: translateY(var(--spacing-1)) !important;
      display: table;
      .bp3-menu {
        max-height: 350px;
        width: 100%;
        overflow: auto;
        border-radius: var(--spacing-2);
        padding: var(--spacing-1) 0px;
        box-sizing: border-box;
        min-width: auto;
        :local {
          .menuItem {
            height: var(--spacing-8);
            padding: 0 var(--spacing-4);
            display: flex;
            align-items: center;
            cursor: pointer;
            p {
              color: var(--black) !important;
              font-weight: 500 !important;
              font-size: 12px !important;
            }
            :global(.bp3-icon) {
              color: var(--grey-900);
              margin-right: var(--spacing-4);
            }
            &.active {
              background: var(--primary-7);
              p {
                color: var(--white) !important;
              }
            }
            &.disabled {
              cursor: not-allowed;
              p {
                color: var(--grey-400) !important;
              }
            }
            &:hover:local:not(.active, .disabled) {
              background: var(--primary-1);
            }
          }
        }
      }
    }
  }
}
