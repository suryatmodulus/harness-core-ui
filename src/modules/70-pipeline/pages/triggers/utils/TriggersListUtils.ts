/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@wings-software/uicore'
// temporary mock data
import { parse } from 'yaml'
import type { AddDrawerMapInterface } from '@common/components/AddDrawer/AddDrawer'
import type { StringKeys } from 'framework/strings'
import { manifestTypeIcons, ManifestDataType } from '@pipeline/components/ManifestSelection/Manifesthelper'
import { ArtifactIconByType, ENABLED_ARTIFACT_TYPES } from '@pipeline/components/ArtifactsSelection/ArtifactHelper'
import { TriggerTypes, AWS_CODECOMMIT, AwsCodeCommit } from './TriggersWizardPageUtils'

export const GitSourceProviders: Record<string, { value: string; iconName: IconName }> = {
  GITHUB: { value: 'Github', iconName: 'github' },
  GITLAB: { value: 'Gitlab', iconName: 'service-gotlab' },
  BITBUCKET: { value: 'Bitbucket', iconName: 'bitbucket-selected' },
  AWS_CODECOMMIT: { value: 'AwsCodeCommit', iconName: 'service-aws-code-deploy' },
  CUSTOM: { value: 'Custom', iconName: 'build' }
}

const TriggerTypeIcons = {
  SCHEDULE: 'trigger-schedule',
  NEW_ARTIFACT: 'new-artifact'
}
export const getTriggerIcon = ({
  type,
  webhookSourceRepo,
  buildType
}: {
  type: string
  webhookSourceRepo?: string // string temporary until backend
  buildType?: string
}): IconName => {
  const updatedWebhookSourceRepo =
    webhookSourceRepo === AwsCodeCommit ? AWS_CODECOMMIT : webhookSourceRepo?.toUpperCase()
  const webhookSourceRepoIconName =
    webhookSourceRepo && updatedWebhookSourceRepo && GitSourceProviders[updatedWebhookSourceRepo]?.iconName
  if (type === TriggerTypes.WEBHOOK && webhookSourceRepoIconName) {
    return webhookSourceRepoIconName as IconName
  } else if (type === TriggerTypes.SCHEDULE) {
    return TriggerTypeIcons.SCHEDULE as IconName
  } else if (type === TriggerTypes.MANIFEST && buildType) {
    if (buildType === ManifestDataType.HelmChart) {
      return manifestTypeIcons.HelmChart
    }
  } else if (type === TriggerTypes.ARTIFACT && buildType) {
    switch (buildType) {
      case ENABLED_ARTIFACT_TYPES.Gcr:
        return ArtifactIconByType.Gcr
      case ENABLED_ARTIFACT_TYPES.Ecr:
        return ArtifactIconByType.Ecr
      case ENABLED_ARTIFACT_TYPES.DockerRegistry:
        return ArtifactIconByType.DockerRegistry
    }
  }
  return 'yaml-builder-trigger'
}

const triggerDrawerMap = (getString: (key: StringKeys) => string): AddDrawerMapInterface => ({
  drawerLabel: getString('pipeline.triggers.triggersLabel'),
  showAllLabel: getString('pipeline.triggers.showAllTriggers'),
  categories: [
    {
      categoryLabel: getString('execution.triggerType.WEBHOOK'),
      categoryValue: 'Webhook',
      items: [
        {
          itemLabel: getString('common.repo_provider.githubLabel'),
          value: GitSourceProviders.GITHUB.value,
          iconName: GitSourceProviders.GITHUB.iconName
        },
        {
          itemLabel: getString('common.repo_provider.gitlabLabel'),
          value: GitSourceProviders.GITLAB.value,
          iconName: GitSourceProviders.GITLAB.iconName
        },
        {
          itemLabel: getString('common.repo_provider.bitbucketLabel'),
          value: GitSourceProviders.BITBUCKET.value,
          iconName: GitSourceProviders.BITBUCKET.iconName
        },
        {
          itemLabel: getString('common.repo_provider.awscodecommit'),
          value: GitSourceProviders.AWS_CODECOMMIT.value,
          iconName: GitSourceProviders.AWS_CODECOMMIT.iconName
        },
        {
          itemLabel: getString('common.repo_provider.customLabel'),
          value: GitSourceProviders.CUSTOM.value,
          iconName: GitSourceProviders.CUSTOM.iconName
        }
      ]
    },
    {
      categoryLabel: getString('pipeline.triggers.scheduledLabel'),
      categoryValue: 'Scheduled',
      items: [
        {
          itemLabel: getString('pipeline.triggers.cronLabel'),
          value: 'Cron',
          iconName: TriggerTypeIcons.SCHEDULE as IconName
        }
      ]
    }
  ]
})

export const getSourceRepoOptions = (getString: (str: StringKeys) => string): { label: string; value: string }[] => [
  { label: getString('common.repo_provider.githubLabel'), value: GitSourceProviders.GITHUB.value },
  { label: getString('common.repo_provider.gitlabLabel'), value: GitSourceProviders.GITLAB.value },
  { label: getString('common.repo_provider.bitbucketLabel'), value: GitSourceProviders.BITBUCKET.value },
  { label: getString('common.repo_provider.codecommit'), value: GitSourceProviders.AWS_CODECOMMIT.value },
  { label: getString('common.repo_provider.customLabel'), value: GitSourceProviders.CUSTOM.value }
]

export const getCategoryItems = (getString: (key: StringKeys) => string): AddDrawerMapInterface =>
  triggerDrawerMap(getString)

export interface ItemInterface {
  itemLabel: string
  iconName: IconName
  value: string
  visible?: boolean
  disabled?: boolean
  categoryValue?: string
}

export interface TriggerDataInterface {
  triggerType: string
  sourceRepo?: string
  manifestType?: string
  artifactType?: string
}

export const getEnabledStatusTriggerValues = ({
  data,
  enabled,
  getString
}: {
  data: any
  enabled: boolean
  getString: (key: StringKeys) => string
}): { values?: any; error?: string } => {
  try {
    const triggerResponseJson = parse(data?.yaml || '')
    triggerResponseJson.trigger.enabled = enabled
    return { values: triggerResponseJson.trigger }
  } catch (e) {
    return { error: getString('pipeline.triggers.cannotParseTriggersData') }
  }
}
const TriggerStatus = {
  FAILED: 'FAILED',
  UNKNOWN: 'UNKNOWN',
  ERROR: 'ERROR',
  TIMEOUT: 'TIMEOUT',
  UNAVAILABLE: 'UNAVAILABLE',
  SUCCESS: 'SUCCESS'
}

export const errorStatusList = [
  TriggerStatus.FAILED,
  TriggerStatus.UNKNOWN,
  TriggerStatus.ERROR,
  TriggerStatus.TIMEOUT,
  TriggerStatus.UNAVAILABLE
]
