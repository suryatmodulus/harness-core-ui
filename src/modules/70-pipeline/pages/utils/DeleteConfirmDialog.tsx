/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState } from 'react'
import { Intent, TextArea } from '@blueprintjs/core'
import { Text, useConfirmationDialog } from '@wings-software/uicore'
import type { EntityGitDetails } from 'services/pipeline-ng'
import { useStrings } from 'framework/strings'
import type { PipelineDTO } from '../pipelines/views/PipelineListView'
import type { InputSetLocal } from '../inputSet-list/InputSetListView'

const useGetEntityText = (entityType: string): string => {
  const { getString } = useStrings()

  switch (entityType) {
    case 'pipeline':
      return getString('common.pipeline')
    case 'inputSet':
      return getString('inputSets.inputSetLabel')
    case 'overlayInputSet':
      return getString('inputSets.overlayInputSet')
    case 'template':
      return getString('common.template.label')
    default:
      return ''
  }
}
interface DeleteConfirmDialogContentProps {
  gitDetails?: EntityGitDetails
  entityName: string
  commitMsg: string
  onCommitMsgChange: (commitMsg: string) => void
  entityType: string
  forceComments?: boolean
}

export const DeleteConfirmDialogContent: React.FC<DeleteConfirmDialogContentProps> = ({
  gitDetails = {},
  entityName = '',
  commitMsg,
  onCommitMsgChange,
  entityType = '',
  forceComments = false
}): JSX.Element => {
  const { getString } = useStrings()
  return (
    <div className={'pipelineDeleteDialog'}>
      <Text margin={{ bottom: 'medium' }} title={entityName}>{`${getString(
        'common.git.confirmDelete'
      )} ${useGetEntityText(entityType)} ${entityName}?`}</Text>
      {(gitDetails?.objectId || forceComments) && (
        <>
          <Text>{forceComments ? getString('common.comments') : getString('common.git.commitMessage')}</Text>
          <TextArea
            value={commitMsg}
            onInput={(event: React.ChangeEvent<HTMLTextAreaElement>) => {
              onCommitMsgChange(event.target.value)
            }}
          />
        </>
      )}
    </div>
  )
}

interface DeleteMetaData {
  versions?: string[]
}

export default function useDeleteConfirmationDialog(
  entityData: PipelineDTO | InputSetLocal,
  entityType: string,
  onDelete?: (commitMsg: string, versions?: string[]) => Promise<void>,
  forceComments = false
): { confirmDelete: (deleteMetaData?: DeleteMetaData) => void } {
  const { getString } = useStrings()
  const [deleteCallMetaData, setDeleteCallMetaData] = useState<DeleteMetaData>()
  const [commitMsg, setCommitMsg] = React.useState<string>(`${getString('delete')} ${entityData.name}`)

  React.useEffect(() => {
    setCommitMsg(`${getString('delete')} ${entityData.name}`)
  }, [entityData])

  const { openDialog: openConfirmDeleteDialog } = useConfirmationDialog({
    contentText: (
      <DeleteConfirmDialogContent
        entityName={entityData?.name || ''}
        entityType={entityType}
        gitDetails={entityData.gitDetails}
        commitMsg={commitMsg}
        onCommitMsgChange={setCommitMsg}
        forceComments={forceComments}
      />
    ),
    titleText: `${getString('delete')} ${useGetEntityText(entityType)}`,
    confirmButtonText: getString('delete'),
    cancelButtonText: getString('cancel'),
    intent: Intent.DANGER,
    buttonIntent: Intent.DANGER,
    onCloseDialog: async (isConfirmed: boolean) => {
      /* istanbul ignore else */
      if (isConfirmed) {
        await onDelete?.(commitMsg, deleteCallMetaData?.versions)
      }
    }
  })

  const confirmDelete = (deleteMetaData?: DeleteMetaData): void => {
    setDeleteCallMetaData(deleteMetaData)
    openConfirmDeleteDialog()
  }
  return { confirmDelete }
}
