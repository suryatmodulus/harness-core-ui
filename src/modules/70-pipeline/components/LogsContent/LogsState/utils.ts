/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { escapeStringRegexp } from '@common/utils/StringUtils'

import type { State } from './types'

export function getRegexForSearch(str: string): RegExp {
  return new RegExp(escapeStringRegexp(str), 'gi')
}

export function getDefaultReducerState(props: Partial<Pick<State, 'selectedStage' | 'selectedStep'>> = {}): State {
  return {
    units: [],
    logKeys: [],
    selectedStage: props.selectedStage || '',
    selectedStep: props.selectedStep || '',
    dataMap: {},
    searchData: {
      text: '',
      currentIndex: 0,
      linesWithResults: []
    }
  }
}

export const logsCache = new Map<string, string>()
