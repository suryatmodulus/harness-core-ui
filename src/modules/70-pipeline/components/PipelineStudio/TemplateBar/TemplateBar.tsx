/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Color, Container, Icon, IconName, Layout, Popover, Text, useConfirmationDialog } from '@wings-software/uicore'
import { Intent, Menu, Position } from '@blueprintjs/core'
import cx from 'classnames'
import { defaultTo } from 'lodash-es'
import { useParams } from 'react-router-dom'
import { useStrings } from 'framework/strings'
import { getTemplateNameWithLabel } from '@pipeline/utils/templateUtils'
import { useFeature } from '@common/hooks/useFeatures'
import { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { useGetTemplate } from 'services/template-ng'
import { getIdentifierFromValue, getScopeFromValue } from '@common/components/EntityReference/EntityReference'
import { Scope } from '@common/interfaces/SecretsInterface'
import type { TemplateLinkConfig } from 'services/pipeline-ng'
import css from './TemplateBar.module.scss'

interface TemplateMenuItem {
  icon?: IconName
  label: string
  disabled?: boolean
  onClick: () => void
}

export interface TemplateBarProps {
  templateLinkConfig: TemplateLinkConfig
  onOpenTemplateSelector: () => void
  onRemoveTemplate: () => Promise<void>
  className?: string
}

export const TemplateBar = (props: TemplateBarProps): JSX.Element => {
  const { templateLinkConfig, onOpenTemplateSelector, onRemoveTemplate, className = '' } = props
  const [menuOpen, setMenuOpen] = React.useState(false)
  const { getString } = useStrings()
  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps>()
  const scope = getScopeFromValue(templateLinkConfig.templateRef)
  const { enabled: templatesEnabled } = useFeature({
    featureRequest: {
      featureName: FeatureIdentifier.TEMPLATE_SERVICE
    }
  })

  const { data } = useGetTemplate({
    templateIdentifier: getIdentifierFromValue(templateLinkConfig.templateRef),
    queryParams: {
      accountIdentifier: accountId,
      projectIdentifier: scope === Scope.PROJECT ? projectIdentifier : undefined,
      orgIdentifier: scope === Scope.PROJECT || scope === Scope.ORG ? orgIdentifier : undefined,
      versionLabel: defaultTo(templateLinkConfig.versionLabel, '')
    }
  })

  const { openDialog: openRemoveTemplateDialog } = useConfirmationDialog({
    intent: Intent.DANGER,
    buttonIntent: Intent.DANGER,
    cancelButtonText: getString('cancel'),
    contentText: getString('pipeline.removeTemplate'),
    titleText: `${getString('common.remove')} ${getTemplateNameWithLabel(data?.data)}?`,
    confirmButtonText: getString('common.remove'),
    onCloseDialog: async isConfirmed => {
      if (isConfirmed) {
        await onRemoveTemplate()
      }
    }
  })

  const getItems = (): TemplateMenuItem[] => {
    return [
      {
        icon: 'command-switch',
        label: getString('pipeline.changeTemplateLabel'),
        disabled: !templatesEnabled,
        onClick: onOpenTemplateSelector
      },
      {
        icon: 'main-trash',
        label: getString('pipeline.removeTemplateLabel'),
        onClick: openRemoveTemplateDialog
      }
    ]
  }

  return (
    <Container
      padding={{ top: 'small', right: 'medium', bottom: 'small', left: 'medium' }}
      background={Color.PRIMARY_6}
      border={{ radius: 4 }}
      className={cx(css.container, className)}
    >
      <Layout.Horizontal spacing={'small'} flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
        <Icon size={11} color={Color.WHITE} name={'template-library'} />
        <Text style={{ flexGrow: 1 }} font={{ size: 'small' }} color={Color.WHITE}>
          {data?.data ? `Using Template: ${getTemplateNameWithLabel(data?.data)}` : getString('loading')}
        </Text>
        <Popover
          isOpen={menuOpen}
          onInteraction={nextOpenState => {
            setMenuOpen(nextOpenState)
          }}
          position={Position.BOTTOM_RIGHT}
          className={css.main}
          disabled={!data?.data}
          portalClassName={css.popover}
        >
          <Icon
            name={'more'}
            color={Color.WHITE}
            className={css.actionButton}
            onClick={e => {
              e.stopPropagation()
              setMenuOpen(true)
            }}
          />
          <Menu style={{ minWidth: 'unset' }} onClick={e => e.stopPropagation()}>
            {getItems()?.map(item => {
              return (
                <li
                  key={item.label}
                  className={cx(css.menuItem, { [css.disabled]: item.disabled })}
                  onClick={e => {
                    e.stopPropagation()
                    if (!item.disabled) {
                      item.onClick()
                      setMenuOpen(false)
                    }
                  }}
                >
                  {item.icon && <Icon name={item.icon} size={12} />}
                  <Text lineClamp={1}>{item.label}</Text>
                </li>
              )
            })}
          </Menu>
        </Popover>
      </Layout.Horizontal>
    </Container>
  )
}
