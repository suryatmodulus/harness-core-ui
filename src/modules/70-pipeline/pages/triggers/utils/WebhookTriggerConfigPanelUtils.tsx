/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { SetStateAction, Dispatch } from 'react'
import cx from 'classnames'
import { FormInput, SelectOption, Text, Container, FontVariation, Color } from '@wings-software/uicore'
import { isEmpty, isUndefined } from 'lodash-es'
import type { StringKeys, UseStringsReturn } from 'framework/strings'
import { GitSourceProviders } from './TriggersListUtils'
import { ConnectorSection } from '../views/ConnectorSection'
import { eventTypes } from './TriggersWizardPageUtils'
import css from '../views/WebhookTriggerConfigPanel.module.scss'

export const autoAbortPreviousExecutionsTypes = [
  eventTypes.PUSH,
  eventTypes.PULL_REQUEST,
  eventTypes.ISSUE_COMMENT,
  eventTypes.MERGE_REQUEST
]

export const getAutoAbortDescription = ({
  event,
  getString
}: {
  event: string
  getString: (key: StringKeys) => string
}): string => {
  if (event === eventTypes.PUSH) {
    return getString('pipeline.triggers.triggerConfigurationPanel.autoAbortPush')
  } else if (event === eventTypes.PULL_REQUEST || event === eventTypes.MERGE_REQUEST) {
    return getString('pipeline.triggers.triggerConfigurationPanel.autoAbortPR')
  } else if (event === eventTypes.ISSUE_COMMENT) {
    return getString('pipeline.triggers.triggerConfigurationPanel.autoAbortIssueComment')
  }
  return ''
}

export const handleSourceRepoChange = ({ e, formikProps }: { e: SelectOption; formikProps: any }): void => {
  if (e.value === GitSourceProviders.CUSTOM.value) {
    formikProps.setValues({
      ...formikProps.values,
      sourceRepo: e.value,
      connectorRef: undefined,
      repoName: '',
      actions: undefined,
      anyAction: false,
      secretToken: undefined
    })
  } else {
    formikProps.setValues({
      ...formikProps.values,
      sourceRepo: e.value,
      connectorRef: undefined,
      repoName: '',
      actions: undefined,
      anyAction: false,
      secretToken: undefined,
      headerConditions: undefined
    })
  }
}

export const renderNonCustomEventFields = ({
  sourceRepo,
  formikProps,
  event,
  eventOptions,
  getString,
  actionsOptions,
  actions
}: {
  sourceRepo?: string
  formikProps: any
  event?: string
  eventOptions: SelectOption[]
  getString: UseStringsReturn['getString']
  actionsOptions: SelectOption[]
  actions: SelectOption[]
}): JSX.Element => {
  return (
    <>
      {sourceRepo && <ConnectorSection formikProps={formikProps} />}
      <FormInput.Select
        className={cx(!event && css.bottomMarginZero)}
        key={event}
        label={getString('pipeline.triggers.triggerConfigurationPanel.event')}
        name="event"
        placeholder={getString('pipeline.triggers.triggerConfigurationPanel.eventPlaceholder')}
        items={eventOptions}
        onChange={e => {
          const additionalValues: any = {}

          if (e.value === eventTypes.PUSH) {
            additionalValues.sourceBranchOperator = undefined
            additionalValues.sourceBranchValue = undefined
          }

          if (e.value !== eventTypes.TAG) {
            additionalValues.tagConditionOperator = undefined
            additionalValues.tagConditionValue = undefined
          }

          formikProps.setValues({
            ...formikProps.values,
            event: e.value,
            actions: e.value === eventTypes.PUSH ? [] : undefined,
            anyAction: false,
            ...additionalValues
          })
        }}
      />
      {event && (
        <>
          {actionsOptions?.length !== 0 && (
            <div className={css.actionsContainer}>
              <div>
                <Text
                  font={{ variation: FontVariation.FORM_INPUT_TEXT, weight: 'semi-bold' }}
                  color={Color.GREY_600}
                  style={{ marginBottom: 'var(--spacing-xsmall)' }}
                >
                  {getString('pipeline.triggers.triggerConfigurationPanel.actions')}
                </Text>
                <FormInput.MultiSelect
                  className={css.multiSelect}
                  name="actions"
                  items={actionsOptions}
                  // yaml design: empty array means selecting all
                  disabled={Array.isArray(actions) && isEmpty(actions)}
                  onChange={e => {
                    if (!e || (Array.isArray(e) && isEmpty(e))) {
                      formikProps.setFieldValue('actions', undefined)
                    } else {
                      formikProps.setFieldValue('actions', e)
                    }
                  }}
                />
              </div>
              <Container className={css.anyActionContainer}>
                <FormInput.CheckBox
                  name="anyAction"
                  key={Date.now()}
                  label={getString('pipeline.triggers.triggerConfigurationPanel.anyActions')}
                  defaultChecked={(Array.isArray(actions) && actions.length === 0) || false}
                  className={css.checkboxAlignment}
                  onClick={(e: React.FormEvent<HTMLInputElement>) => {
                    formikProps.setFieldTouched('actions', true)
                    if (e.currentTarget?.checked) {
                      formikProps.setFieldValue('actions', [])
                    } else {
                      formikProps.setFieldValue('actions', undefined)
                    }
                  }}
                />
              </Container>
            </div>
          )}
          {autoAbortPreviousExecutionsTypes.includes(event) && (
            <>
              <FormInput.CheckBox
                style={{ position: 'relative', left: '0' }}
                name="autoAbortPreviousExecutions"
                label="Auto-abort Previous Execution"
                className={css.checkboxAlignment}
              />
              <Text className={css.autoAbortDescription}>{getAutoAbortDescription({ event, getString })}</Text>
            </>
          )}
        </>
      )}
    </>
  )
}
interface ActionsOptionsMapInterface {
  [key: string]: string[]
}

export const getEventLabelMap = (event: string): string => {
  // add space between camelcase-separated words
  if (event === eventTypes.PULL_REQUEST) {
    return 'Pull Request'
  } else if (event === eventTypes.MERGE_REQUEST) {
    return 'Merge Request'
  } else if (event === eventTypes.ISSUE_COMMENT) {
    return 'Issue Comment'
  }
  return event
}

export const getEventAndActions = ({
  data,
  sourceRepo
}: {
  data: {
    [key: string]: {
      [key: string]: string[]
    }
  }
  sourceRepo: string
}): { eventOptions: SelectOption[]; actionsOptionsMap: ActionsOptionsMapInterface } => {
  const filteredData = data?.[sourceRepo] || {}
  const eventOptions = Object.keys(filteredData).map(event => ({
    label: getEventLabelMap(event),
    value: event
  }))
  return { eventOptions, actionsOptionsMap: filteredData }
}

export const clearEventsAndActions = ({
  newActionsOptionsMap,
  formikProps,
  setActionsOptions,
  event
}: {
  newActionsOptionsMap: ActionsOptionsMapInterface
  formikProps: any
  setActionsOptions: Dispatch<SetStateAction<SelectOption[]>>
  event: string
}): void => {
  if (isUndefined(newActionsOptionsMap[event])) {
    formikProps.setFieldValue('event', undefined)
    return
  }

  const newActionsOptions = newActionsOptionsMap[event]?.map((val: string) => ({ label: val, value: val }))
  setActionsOptions(newActionsOptions)
  if (newActionsOptions?.length === 0) {
    formikProps.setFieldValue('actions', [])
  }
}
