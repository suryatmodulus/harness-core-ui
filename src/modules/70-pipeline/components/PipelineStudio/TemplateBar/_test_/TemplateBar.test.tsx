/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, getByText, render } from '@testing-library/react'
import { findDialogContainer, findPopoverContainer, TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import { accountPathProps, pipelineModuleParams, pipelinePathProps } from '@common/utils/routeUtils'
import { TemplateBar } from '../TemplateBar'

jest.mock('@pipeline/utils/useStageTemplateActions', () => ({
  useStageTemplateActions: jest.fn().mockReturnValue({
    onUseTemplate: jest.fn(),
    onCopyTemplate: jest.fn(),
    onRemoveTemplate: jest.fn(),
    onOpenTemplateSelector: jest.fn()
  })
}))

jest.mock('services/template-ng', () => ({
  useGetTemplate: jest.fn().mockImplementation(() => ({
    status: 'SUCCESS',
    data: {
      data: {
        name: 'New Stage Name',
        identifier: 'new_stage_name',
        versionLabel: 'v1'
      }
    }
  }))
}))

describe('<TemplateBar /> tests', () => {
  test('should match snapshot and work as expected', async () => {
    const props = {
      templateLinkConfig: {
        templateRef: 'New_CD_Stage_Name',
        versionLabel: 'Version1'
      },
      onOpenTemplateSelector: jest.fn(),
      onRemoveTemplate: jest.fn()
    }
    const { container } = render(
      <TestWrapper
        path={routes.toPipelineStudio({ ...accountPathProps, ...pipelinePathProps, ...pipelineModuleParams })}
        pathParams={{
          pipelineIdentifier: 'stage1',
          accountId: 'accountId',
          orgIdentifier: 'CV',
          projectIdentifier: 'Milos2',
          module: 'cd'
        }}
      >
        <TemplateBar {...props} />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
    const optionsBtn = container.querySelector('.bp3-icon-more') as HTMLElement
    await act(async () => {
      fireEvent.click(optionsBtn)
    })
    const popover = findPopoverContainer()
    const changeBtn = getByText(popover as HTMLElement, 'pipeline.changeTemplateLabel')
    await act(async () => {
      fireEvent.click(changeBtn)
    })
    expect(props.onOpenTemplateSelector).toBeCalled()
    const removeBtn = getByText(popover as HTMLElement, 'pipeline.removeTemplateLabel')
    await act(async () => {
      fireEvent.click(removeBtn)
    })
    const submitBtn = getByText(findDialogContainer() as HTMLElement, 'common.remove')
    await act(async () => {
      fireEvent.click(submitBtn)
    })
    expect(props.onRemoveTemplate).toBeCalled()
  })
})
