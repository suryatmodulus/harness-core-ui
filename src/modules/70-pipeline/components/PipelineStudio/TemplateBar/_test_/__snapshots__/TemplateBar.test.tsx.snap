// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateBar /> tests should match snapshot and work as expected 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main container StyledProps--padding-top-small StyledProps--padding-right-medium StyledProps--padding-bottom-small StyledProps--padding-left-medium StyledProps--background-primary6 StyledProps--border"
  >
    <div
      class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-white"
        data-icon="template-library"
      >
        <svg
          height="11"
          viewBox="0 0 12 13"
          width="11"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs>
            <path
              d="M1.8 13c.2 0 .4-.3.4-.6V.6C2.2.3 2 0 1.8 0H.4C.2 0 0 .3 0 .6v11.7c0 .4.2.7.4.7h1.4zm8.8 0l.9-.3c.3-.1.5-.4.4-.8L8.9.6C8.8.3 8.5.1 8.1.2l-.9.3c-.4 0-.5.4-.5.7l3.1 11.3c.1.4.5.6.8.5zm-5 0c.2 0 .4-.3.4-.7V.6c0-.3-.2-.6-.4-.6H4.2c-.2 0-.4.3-.4.6v11.7c0 .4.2.7.4.7h1.4z"
              id="template-library_svg__a"
            />
          </defs>
          <clippath
            id="template-library_svg__b"
          >
            <use
              overflow="visible"
              xlink:href="#template-library_svg__a"
            />
          </clippath>
          <path
            clip-path="url(#template-library_svg__b)"
            d="M-5-5h22v23H-5z"
            fill="#25a6f7"
          />
        </svg>
      </span>
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-white"
        style="flex-grow: 1;"
      >
        Using Template: New Stage Name (v1)
      </p>
      <span
        class="bp3-popover-wrapper main"
      >
        <span
          class="bp3-popover-target"
        >
          <span
            class="bp3-icon bp3-icon-more StyledProps--main actionButton StyledProps--color StyledProps--color-white"
            icon="more"
          >
            <svg
              data-icon="more"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                more
              </desc>
              <path
                d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;
